src/storage/mvcc/reader/forward_scanner.rs

```diff
@@ -0,0 +1,763 @@
+
+            if has_lock {
+                match self.isolation_level {
+                    IsolationLevel::SI => {
+                        // Only needs to check lock in SI
+                        match super::util::load_and_check_lock_from_cursor(
+                            &mut self.lock_cursor,
+                            &current_user_key,
+                            self.ts,
+                            &mut self.statistics,
+                        )? {
+                            CheckLockResult::NotLocked => {}
+                            CheckLockResult::Locked(e) => result = Err(e),
+                            CheckLockResult::Ignored(ts) => get_ts = ts,
+                        }
+                    }
+                    IsolationLevel::RC => {}
+                }
+                self.lock_cursor.next(&mut self.statistics.lock);
+            }
+            if has_write {
+                // We don't need to read version if there is a lock error already.
+                if result.is_ok() {
+                    // Attempt to read specified version of the key. Note that we may get `None`
+                    // indicating that no desired version is found, or a DELETE version is found
+                    result = self.get(&current_user_key, get_ts, &mut met_next_user_key);
+                }
+                // Even if there is a lock error, we still need to step the cursor for future
+                // calls. However if we are already pointing at next user key, we don't need to
+                // move it any more. `met_next_user_key` eliminates a key compare.
+                if !met_next_user_key {
+                    self.move_write_cursor_to_next_user_key(&current_user_key)?;
+                }
+            }
+
+            // If we got something, it can be just used as the return value. Otherwise, we need
+            // to continue stepping the cursor.
+            if let Some(v) = result? {
+                return Ok(Some((current_user_key, v)));
+            }
+        }
+    }
+
+    /// Attempt to get the value of a key specified by `user_key` and `self.ts`. This function
+    /// requires that the write cursor is currently pointing to the latest version of `user_key`.
+    #[inline]
+    fn get(
+        &mut self,
+        user_key: &Key,
+        ts: u64,
+        met_next_user_key: &mut bool,
+    ) -> Result<Option<Value>> {
+        assert!(self.write_cursor.valid());
+
+        // The logic starting from here is similar to `PointGetter`.
+
+        // Try to iterate to `${user_key}_${ts}`. We first `next()` for a few times,
+        // and if we have not reached where we want, we use `seek()`.
+
+        // Whether we have *not* reached where we want by `next()`.
+        let mut needs_seek = true;
+
+        for i in 0..SEEK_BOUND {
+            if i > 0 {
+                self.write_cursor.next(&mut self.statistics.write);
+                if !self.write_cursor.valid() {
```

Comment:
`next()` will return whether valid or not

Reply:
According to siddontang's suggestion, we should only use `valid()` to judge whether cursor is valid or not, and avoid using the return value. Return values will be removed after the refinement.